openstack/identity/v3/extensions/ec2tokens/requests.go

```diff
@@ -0,0 +1,279 @@
+		return ""
+	}
+	return canonicalQsV2(params)
+}
+
+func canonicalHeadersV4(headers map[string]string, signedHeaders string) string {
+	headersLower := make(map[string]string, len(headers))
+	for k, v := range headers {
+		headersLower[strings.ToLower(k)] = v
+	}
+
+	var headersList []string
+	for _, h := range strings.Split(signedHeaders, ";") {
+		if v, ok := headersLower[h]; ok {
+			headersList = append(headersList, h+":"+v)
+		}
+	}
+
+	return strings.Join(headersList, "\n") + "\n"
+}
+
+func sumHMAC1(key []byte, data []byte) []byte {
+	hash := hmac.New(sha1.New, key)
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
+func sumHMAC256(key []byte, data []byte) []byte {
+	hash := hmac.New(sha256.New, key)
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
+func signatureKeyV4(secret, date, region, service string) []byte {
+	kDate := sumHMAC256([]byte("AWS4"+secret), []byte(date))
+	kRegion := sumHMAC256(kDate, []byte(region))
+	kService := sumHMAC256(kRegion, []byte(service))
+	return sumHMAC256(kService, []byte(aws4Request))
+}
+
+func signatureV4(opts AuthOptions, params map[string]string, date time.Time, bodyHash string) string {
+	scope := strings.Join([]string{
+		date.Format(yyyymmdd),
+		opts.Region,
+		opts.Service,
+		aws4Request,
+	}, "/")
+
+	canonicalRequest := strings.Join([]string{
+		opts.Verb,
+		opts.Path,
+		canonicalQsV4(opts.Verb, params),
+		canonicalHeadersV4(opts.Headers, params["X-Amz-SignedHeaders"]),
+		params["X-Amz-SignedHeaders"],
+		bodyHash,
+	}, "\n")
+	hash := sha256.Sum256([]byte(canonicalRequest))
+
+	strToSign := strings.Join([]string{
+		v4Algo,
+		date.Format(iso8601utc),
+		scope,
+		hex.EncodeToString(hash[:]),
+	}, "\n")
+
+	key := signatureKeyV4(opts.Secret, date.Format(yyyymmdd), opts.Region, opts.Service)
+
+	return hex.EncodeToString(sumHMAC256(key, []byte(strToSign)))
+}
+
+func randomBodyHash() string {
+	h := make([]byte, 64)
+	rand.Read(h)
+	return hex.EncodeToString(h)
+}
+
+func paramsToMap(c map[string]interface{}) map[string]string {
+	// convert map[string]interface{} to map[string]string
+	p := make(map[string]string)
+	if v, _ := c["params"].(map[string]interface{}); v != nil {
+		for k, v := range v {
+			p[k] = v.(string)
+		}
+	}
+
+	c["params"] = p
+
+	return p
+}
+
+// ToTokenV3CreateMap is a dummy method to satisfy tokens.AuthOptionsBuilder interface
+func (opts *AuthOptions) ToTokenV3ScopeMap() (map[string]interface{}, error) {
+	return nil, nil
+}
+
+// CanReauth is a method method to satisfy tokens.AuthOptionsBuilder interface
+func (opts *AuthOptions) CanReauth() bool {
+	return opts.AllowReauth
+}
+
+// ToAuthOptionsCreateMap formats an AuthOptions into a create request.
+func (opts *AuthOptions) ToTokenV3CreateMap(map[string]interface{}) (map[string]interface{}, error) {
+	b, err := gophercloud.BuildRequestBody(opts, "credentials")
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Signature != nil {
+		return b, nil
+	}
+
+	// calculate signature, when it is not set
```

Comment:
Is there any benefit to refactoring this `ToTokenV3CreateMap` method to look more like this?

https://github.com/openstack/python-keystoneclient/blob/526350eaeeadef8c1ca5f80547c01f94342a6653/keystoneclient/contrib/ec2/utils.py#L87-L104

Reply:
my aim was to reduce an amount of if/then/else. added more comments.